elastic/elasticsearch
# [Fleet] Make `.fleet-secrets` index internal

### Context

Fleet team is currently implementing phase 1 of storing Fleet secrets ([meta issue](https://github.com/elastic/ingest-dev/issues/1000), [phase 1 issue](https://github.com/elastic/kibana/issues/154715)), where a a new system index was created to store the secrets (cf. [this schema](https://user-images.githubusercontent.com/1813008/218431667-079a6ca8-b046-4a88-b085-4ac522c33e9e.png)).

The access expectations are as follows:
   - Kibana system should only have write and delete access, NOT read access.
   - Fleet server should have read access only.

The `.fleet-secrets` system index was added in https://github.com/elastic/elasticsearch/pull/95625; delete privilege was added in https://github.com/elastic/elasticsearch/pull/96080.

Upon testing, Mark found out that the `kibana_system` user has all privileges on `.fleet-*` indices, thus breaking the requirements; the initial proposed fixed was to rename the index to `.secrets-fleet` in https://github.com/elastic/elasticsearch/pull/96172. Following discussion on this PR, an alternative approach was decided that will allow better clarity on the different Fleet-related system indices as well as introducing fine-grained access to `.fleet-secrets`:
1. Break down access rules for each individual index (cf. [this comment](https://github.com/elastic/elasticsearch/pull/96172#issuecomment-1551180846)). This change is implemented in https://github.com/elastic/elasticsearch/pull/97881.
2. Make the `.fleet-secrets` index internal (cf. [this comment](https://github.com/elastic/elasticsearch/pull/96172#issuecomment-1553554429)). This is what this PR does.

### What this PR does

* Make the `.fleet-secrets` system index internal.
* Implement the write, delete and read API endpoints.
* Define new cluster privileges to enforce read-only access for Kibana user and write/delete-only access for Fleet server.

### How to test this PR

The API endpoints can be tested in Kibana Dev Tools:

1. Run ES from source with `./gradlew run`.
2. Once ES is running, change the default password for the elastic user from "password" to "changeme" to match Kibana's:
   ```
   curl -k -u elastic-admin:elastic-password -H 'Content-Type: application/json' \
   http://localhost:9200/_security/user/elastic/_password -d'{"password": "changeme"}'
   curl -k -u elastic:changeme -H 'Content-Type: application/json' \
    http://localhost:9200/_security/user/kibana_system/_password -d'{"password": "changeme"}'
   ```
3. Run Kibana with `yarn start`.
4. Log into Kibana and go to `/app/dev_tools#/console`.
5. Create a secret:
   ```
   POST /_fleet/secret/
   {
     "value": "test123"
   }
   ```
6. This should succeed and return the id of the created secret, e.g.:
   ```
   {
     "id": "JoCwlokBZNKF_JG5gPBw"
   }
   ```
7. Copy the id and try to get the secret:
   ```
   GET /_fleet/secret/JoCwlokBZNKF_JG5gPBw
   ```
8. This should succeed and return the secret id and value:
   ```
   {
     "id": "JoCwlokBZNKF_JG5gPBw",
     "value": "test123"
   }
   ```
9. Try to delete the secret:
   ```
   DELETE /_fleet/secret/JoCwlokBZNKF_JG5gPBw
   ```
10. This should succeed and return the following payload:
   ```
   {
     "deleted": true
   }
   ```
11. Try to delete the same secret again. It should fail with 404 and return an error:
   ```
   {
     "error": {
       "root_cause": [
         {
           "type": "resource_not_found_exception",
           "reason": "No secret with id [JoCwlokBZNKF_JG5gPBw]"
         }
       ],
       "type": "resource_not_found_exception",
       "reason": "No secret with id [JoCwlokBZNKF_JG5gPBw]"
     },
     "status": 404
   }
   ```
12. Try to get the same secret again. It should fail with 404 and return the same error as above.
